### PR TITLE
feat(zk): implement faster pke proof

### PR DESCRIPTION
**Pull request by sarah-ek**
### PR content/description
implements Benoit's new private key encryption zk proof

### Check-list:

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [x] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by IceTDrinker** [comment]
cargo builds is red, could be good to have a review with JB potentially

File: tfhe-zk-pok/src/proofs/mod.rs [hunk 1]
```diff
@@ -141,5 +141,12 @@ pub const HASH_METADATA_LEN_BYTES: usize = 256;
 pub mod binary;
 pub mod index;
 pub mod pke;
+pub mod pke_v2 {
+    pub use super::pke_v2_impl::*;
+
+    pub use super::pke_v2_impl::crs_gen_cs as crs_gen;
+}
```
Comment (IceTDrinker) [inline tfhe-zk-pok/src/proofs/mod.rs:148]: what is the reason for the indirection here from pke_v2 to pke_v2_impl ?

File: tfhe-zk-pok/Cargo.toml [hunk 2]
```diff
@@ -22,6 +22,7 @@ rayon = "1.8.0"
 sha3 = "0.10.8"
 serde = { version = "~1.0", features = ["derive"] }
 zeroize = "1.7.0"
+num-bigint = "0.4.5"
```
Comment (IceTDrinker) [inline tfhe-zk-pok/Cargo.toml:25]: in the end do you think it's better to pull that dependency ?

File: tfhe-zk-pok/src/curve_api.rs [hunk 3]
```diff
@@ -329,12 +342,24 @@ impl CurveGroupOps<bls12_446::Zp> for bls12_446::G1 {
...
-        // Self::Affine::multi_mul_scalar(bases, scalars)
+        // overhead seems to not be worth it outside of wasm
+        #[cfg(target_family = "wasm")]
+        {
+            msm::msm_wnaf_g1_446(bases, scalars)
+        }
+        #[cfg(not(target_family = "wasm"))]
+        {
+            Self::Affine::multi_mul_scalar(bases, scalars)
+        }
```
Comment (IceTDrinker) [inline tfhe-zk-pok/src/curve_api.rs:362]: is the `if cfg!(target_family = "wasm") {` usable here ? I think it makes it easier to read

File: tfhe-zk-pok/src/curve_api.rs [hunk 4]
```diff
@@ -365,9 +390,14 @@ impl CurveGroupOps<bls12_446::Zp> for bls12_446::G2 {
...
-        self.mul_scalar(scalar)
+        if scalar.inner == MontFp!("2") {
+            self.double()
+        } else {
+            self.mul_scalar(scalar)
+        }
     }
 
+    #[track_caller]
     fn multi_mul_scalar(bases: &[Self::Affine], scalars: &[bls12_446::Zp]) -> Self {
```
Comment (IceTDrinker) [inline tfhe-zk-pok/src/curve_api.rs:401]: this function can fail ?

Comment (IceTDrinker) [reply tfhe-zk-pok/src/curve_api.rs:401]: same question elsewhere

File: tfhe-zk-pok/src/proofs/mod.rs (hunk 1 above)
Comment (sarah-ek) [reply tfhe-zk-pok/src/proofs/mod.rs:148]: just cause i wanted a quick way to swap out pke for pke_v2 for benchmarks. i'll fix it

File: tfhe-zk-pok/Cargo.toml (hunk 2 above)
Comment (sarah-ek) [reply tfhe-zk-pok/Cargo.toml:25]: yeah, doesn't seem to add any compile-time overhead and we can replace it if we ever need to in the future

File: tfhe-zk-pok/src/curve_api.rs (hunk 3 above)
Comment (sarah-ek) [reply tfhe-zk-pok/src/curve_api.rs:362]: yeah

File: tfhe-zk-pok/src/proofs/mod.rs (hunk 1 above)
Comment (IceTDrinker) [reply tfhe-zk-pok/src/proofs/mod.rs:148]: got it

File: tfhe-zk-pok/src/curve_api.rs (hunk 4 above)
Comment (sarah-ek) [reply tfhe-zk-pok/src/curve_api.rs:401]: yeah, if the slices aren't the same size

File: tfhe-zk-pok/src/curve_api/msm.rs [hunk 5]
```diff
@@ -236,207 +237,3 @@ pub fn msm_wnaf_g1_446(
                 total
             })
 }
-
-// Compute msm using windowed non-adjacent form
-pub fn msm_wnaf_g1_446_extended(
```
[the remark this answers is not in the document]
Comment (sarah-ek) [reply tfhe-zk-pok/src/curve_api/msm.rs:241]: this wasn't being used. i only implemented it to experiment with splitting the exponents but i didn't see any gain from it

File: tfhe-zk-pok/src/proofs/mod.rs (hunk 1 above)
Comment (sarah-ek) [reply tfhe-zk-pok/src/proofs/mod.rs]: i removed the indirection

File: tfhe-zk-pok/src/curve_api.rs (hunk 3 above)
Comment (sarah-ek) [reply tfhe-zk-pok/src/curve_api.rs]: fixed

**Comment by IceTDrinker** [comment]
this will need a rebase to have proper CI

File: tfhe-zk-pok/src/curve_api/msm.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe-zk-pok/src/curve_api/msm.rs:241]: got it thanks

File: tfhe-zk-pok/src/proofs/pke_v2.rs [hunk 6]
```diff
@@ -0,0 +1,2194 @@
...
+
+pub fn crs_gen<G: Curve>(
+    d: usize,
+    k: usize,
+    B: u64,
+    q: u64,
+    t: u64,
+    rng: &mut dyn RngCore,
+) -> PublicParams<G> {
+    crs_gen_cs(d, k, B, q, t, rng)
```
Comment (IceTDrinker) [inline tfhe-zk-pok/src/proofs/pke_v2.rs:258]: what's the cs vs. ghl notation ? It seems the CRS we use in the test is the ghl one ? @sarah-ek

**Review by jborfila — APPROVED** [review]

**Pull request merged**